neuroanalysis/event_detection.py

```diff
@@ -95,45 +95,175 @@ def zero_crossing_events(data, min_length=3, min_peak=0.0, min_sum=0.0, noise_th
+        index: int
+           index of the initial crossing of the *threshold*
+        len: int
+            index length of the event
+        sum: float
+           sum of the values in the array between the start and end of the event
+        peak: float
+           peak value of event
+        peak_index: int
+           index value of the peak
+        time: float, or np.nan if timing not available
+           time of the onset of an event
+        duration: float, or np.nan if timing not available
+           duration of time of the event
+        area: float, or np.nan if timing not available
+           area under the curve of the event
+        peak_time: float, or np.nan if timing not available
+           time of peak
     """
-    threshold = abs(threshold)
+
+
     data = trace.data
     data1 = data - baseline
+    # convert threshold array
+    if isinstance(threshold, float):
+        threshold = np.ones(len(data)) * abs(threshold)
+
+
     #if (hasattr(data, 'implements') and data.implements('MetaArray')):
     
-    ## find all threshold crossings
-    masks = [(data1 > threshold).astype(np.byte), (data1 < -threshold).astype(np.byte)]
+    ## find all positive and negative threshold crossings of baseline adjusted data
+
+    if debug:
+        import pdb; pdb.set_trace()
+        # FYI: can't use matplot lib before debugger is on
+        # type *continue to see plot
+        import matplotlib.pyplot as mpl
+
+
+    masks = [(data1 > threshold).astype(np.byte), (data1 < -threshold).astype(np.byte)] # 1 where data is [above threshold, below negative threshold]
     hits = []
     for mask in masks:
-        diff = mask[1:] - mask[:-1]
-        on_inds = list(np.argwhere(diff==1)[:,0] + 1)
-        off_inds = list(np.argwhere(diff==-1)[:,0] + 1)
-        if len(on_inds) == 0 or len(off_inds) == 0:
-            continue
-        if off_inds[0] < on_inds[0]:
-            if omit_ends:
-                off_inds = off_inds[1:]
-                if len(off_inds) == 0:
-                    continue
-            else:
-                on_inds.insert(0, 0)
-        if off_inds[-1] < on_inds[-1]:
-            if omit_ends:
-                on_inds = on_inds[:-1]
-            else:
-                off_inds.append(len(diff))
+        diff = mask[1:] - mask[:-1] # -1 (or +1) when crosses from above to below threshold (or visa versa if threshold is negative). Note above threshold refers to value furthest from zero, i.e. it can be positive or negative
+        
+        # TODO: It might be a good idea to make offindexes one less so that region above threshold looks symmetrical
+        # find start and end inicies (above threshold) where waveform is above threshold
+        on_inds = list(np.argwhere(diff==1)[:,0] + 1) #where crosses from below to above threshold Note taking index after this happens
+        off_inds = list(np.argwhere(diff==-1)[:,0]) #where crosses from above to below threshold. Note taking index before this happens
```

Comment:
Why did you remove the `+1` here? The idea is you could take `data[on_ind:off_ind]` and get exactly the region above threshold.

Reply:
The last index was actually below the threshold and I only wanted to use indexes above threshold.  The +1 includes an area and a value below threshold in the sum and the area.

Reply:
I will also remove the #TODO on 188 as I did implement that here.